hypershift-operator/controllers/nodepool/capi.go

```diff
@@ -0,0 +1,1340 @@
+			return nil
+		}
+		return fmt.Errorf("error getting MachineSet: %w", err)
+	}
+	if ms.DeletionTimestamp != nil {
+		return nil
+	}
+	err = c.Delete(ctx, ms)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("error deleting MachineSet: %w", err)
+	}
+	return nil
+}
+
+func (c *CAPI) Pause(ctx context.Context) error {
+	// Pause MachineSet
+	if err := c.pauseMachineSet(ctx); err != nil {
+		return fmt.Errorf("error pausing MachineSet: %w", err)
+	}
+
+	// Pause MachineDeployment
+	if err := c.pauseMachineDeployment(ctx); err != nil {
+		return fmt.Errorf("error pausing MachineDeployment: %w", err)
+	}
+
+	return nil
+}
+
+func (c *CAPI) pauseMachineSet(ctx context.Context) error {
+	ms := c.machineSet()
+	err := c.Get(ctx, client.ObjectKeyFromObject(ms), ms)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("error getting MachineSet: %w", err)
+	}
+	if ms.Annotations == nil {
+		ms.Annotations = make(map[string]string)
+	}
+	// FIXME: In future we may want to use the spec field instead
+	// https://github.com/kubernetes-sigs/cluster-api/issues/6966
+	// TODO: Also for paused to be complete we will need to pause all MHC if autorepair
+	// is enabled and remove the autoscaling labels from the MachineDeployment / Machineset
+	ms.Annotations[capiv1.PausedAnnotation] = "true"
+	return c.Update(ctx, ms)
+}
+
+func deleteMachineHealthCheck(ctx context.Context, c client.Client, mhc *capiv1.MachineHealthCheck) error {
+	err := c.Get(ctx, client.ObjectKeyFromObject(mhc), mhc)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("error getting MachineHealthCheck: %w", err)
+	}
+	if mhc.DeletionTimestamp != nil {
+		return nil
+	}
+	err = c.Delete(ctx, mhc)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("error deleting MachineHealthCheck: %w", err)
+	}
+	return nil
+}
+
+func (c *CAPI) reconcileMachineDeployment(ctx context.Context, log logr.Logger,
+	machineDeployment *capiv1.MachineDeployment,
+	machineTemplateCR client.Object) error {
+
+	nodePool := c.nodePool
+	capiClusterName := c.capiClusterName
+	// Set annotations and labels
+	if machineDeployment.GetAnnotations() == nil {
+		machineDeployment.Annotations = map[string]string{}
+	}
+	machineDeployment.Annotations[nodePoolAnnotation] = client.ObjectKeyFromObject(nodePool).String()
+	// Delete any paused annotation
+	delete(machineDeployment.Annotations, capiv1.PausedAnnotation)
+	if machineDeployment.GetLabels() == nil {
+		machineDeployment.Labels = map[string]string{}
+	}
+	machineDeployment.Labels[capiv1.ClusterNameLabel] = capiClusterName
+
+	// Set defaults. These are normally set by the CAPI machinedeployment webhook.
+	// However, since we don't run the webhook, CAPI updates the machinedeployment
+	// after it has been created with defaults.
+	machineDeployment.Spec.MinReadySeconds = k8sutilspointer.Int32(0)
```

Comment:
k8sutilspointer.Int32 is deprecated. These should be updated to use ptr.To[int32]